### PR TITLE
build: Add contracts compile and deploy job

### LINKED ISSUE #92: [core] Continuous deployment for smart contracts

### PR DESCRIPTION
**Pull request by bgins**
### Review Type Requested (choose one):

- [x] **Logic** - thorough check (from everybody doing review)

### Summary

This pull request adds a GH action job to compile and deploy smart contracts to devnet.

### Task/Issue reference

Closes: #92 

### Details (optional)

This pull request implements the following changes:

- [x] Add job to deploy contracts to devnet
- [x] Reduce the use of private keys when deploying contracts (only the admin key is needed)
- [x] Read RPC endpoint from environment

Note that this version of the job will deploy every time. We will want to check if re-deployment is necessary in a future PR.

We'll test this on `main` when we merge in #111, but a test run on this branch can be seen here: https://github.com/Lilypad-Tech/lilypad/actions/runs/9212958660/job/25345761697

### CODE REVIEW

File: .github/workflows/devnet_deploy.yml [hunk 1]
```diff
@@ -7,19 +7,25 @@ on:
...
+      - name: Deploy contracts
+        id: deploy-contracts
         env:
-          WEB3_RPC_URL: ${{ secrets.WEB3_RPC_URL }}
+          DOPPLER_TOKEN: ${{ secrets.DOPPLER_TOKEN_CONTRACTS_DEPLOY_DEVNET }}
         run: |
-          echo "# TODO"
+          cd hardhat
+          npm ci
+          doppler run -- npx hardhat deploy --network geth
```
Comment (bgins) [inline .github/workflows/devnet_deploy.yml:28]: Deploy implicitly compiles the contracts.

File: hardhat/hardhat.config.ts [hunk 2]
```diff
@@ -27,16 +27,16 @@ const config: HardhatUserConfig = {
     geth: {
       url: 'http://localhost:8545',
       chainId: 1337,
-      accounts: PRIVATE_KEYS,
+      accounts: [getAccount('admin').privateKey]
```
Comment (AquiGorka) [inline hardhat/hardhat.config.ts:31]: I have my suspicions that these may affect the development setup of the blockchain node.

Could you clear, run, fund and boot to verify this did not impact anything there?

File: hardhat/hardhat.config.ts [hunk 3]
```diff
@@ -15,6 +15,7 @@ const ENV_FILE = process.env.DOTENV_CONFIG_PATH || '../.env'
 dotenv.config({ path: ENV_FILE })
 
 const NETWORK = process.env.NETWORK || "geth";
+const NETWORK_URL = process.env.WEB3_RPC_URL || 'http://localhost:8545';
```
Comment (AquiGorka) [inline hardhat/hardhat.config.ts]: Can we give this a different name? From the way the env var `WEB3_RPC_URL` is used all around it points to the websocket endpoint (which for dev is `localhost:8546`), maybe `WEB3_HTTP_URL`?

Comment (bgins) [reply hardhat/hardhat.config.ts]: Yeah good point. Updated it to `WEB3_HTTP_URL`.

File: hardhat/hardhat.config.ts (hunk 2 above)
Comment (bgins) [reply hardhat/hardhat.config.ts:31]: Just checked and run, fund, and boot for the chain all work with this change in local development.

The reason I think this is safe is the deploy scripts (https://github.com/Lilypad-Tech/lilypad/tree/main/hardhat/deploy) only deploy and excute commands from `admin`.

The public addresses for the other users appear to be configured here:

https://github.com/Lilypad-Tech/lilypad/blob/afbf4f8f9e45acca16615ffe730f19e2dfad2553/hardhat/hardhat.config.ts#L24

And they are referenced, but admin does all the work.

Comment (AquiGorka) [reply hardhat/hardhat.config.ts:31]: Thank you for verifying.

**Review by AquiGorka — APPROVED** [review]
🎉 🎉  thank you @bgins !

**Pull request merged**